File: frontend/src/types/device.ts
// ── Base Response Types ──────────────────────────────────────────────────

export interface BaseResponse<T> {
  errorCode: string
  errorMsg: string
  errorDetail?: string
  timestamp: number
  body: T
}

export interface PageResponse<T> {
  content: T[]
  totalElements: number
  totalPages: number
  page: number
  size: number
}

// ── Device ───────────────────────────────────────────────────────────────

export interface DeviceRequest {
  deviceType: string
  deviceCode: string
  deviceName?: string
  twd97X?: number | null
  twd97Y?: number | null
  lng?: number | null
  lat?: number | null
  elevation?: number | null
  deptId?: number | null
  contractId?: number | null
  propertyOwner?: string
  installedAt?: string | null
  parentDeviceId?: number | null
  mountPosition?: string
  connectivityType?: string
  circuitId?: number | null
  attributes?: Record<string, unknown>
}

export interface DeviceResponse {
  id: number
  deviceType: string
  deviceCode: string
  deviceName?: string
  twd97X?: number | null
  twd97Y?: number | null
  lng?: number | null
  lat?: number | null
  elevation?: number | null
  deptId?: number | null
  deptName?: string
  contractId?: number | null
  contractCode?: string
  propertyOwner?: string
  status: string
  installedAt?: string | null
  decommissionedAt?: string | null
  parentDeviceId?: number | null
  parentDeviceCode?: string
  mountPosition?: string
  connectivityType?: string
  lastHeartbeatAt?: string | null
  circuitId?: number | null
  circuitNumber?: string
  childrenCount: number
  children?: DeviceResponse[]
  createdBy?: string
  createdAt?: string
  updatedAt?: string
}

export interface DeviceStatsResponse {
  totalDevices: number
  byType: Record<string, number>
  byStatus: Record<string, number>
  onlineRate: number
  openFaults: number
}

// ── Work Order ───────────────────────────────────────────────────────────

export interface WorkOrderRequest {
  deviceId?: number | null
  deviceCode?: string
  circuitId?: number | null
  orderType: string
  sourceType: string
  priority?: string
  reporterName?: string
  reporterContact?: string
  description?: string
}

export interface WorkOrderLogEntry {
  action: string
  operatorName?: string
  latitude?: number | null
  longitude?: number | null
  note?: string
  createdAt: string
}

export interface WorkOrderResponse {
  id: number
  deviceId?: number | null
  deviceCode?: string
  deviceName?: string
  circuitId?: number | null
  orderType: string
  sourceType: string
  status: string
  priority?: string
  reporterName?: string
  reporterContact?: string
  reportedAt?: string
  description?: string
  assignedTo?: string
  assignedToName?: string
  assignedAt?: string
  startedAt?: string
  completedAt?: string
  completionRemark?: string
  faultCause?: string
  repairCost?: number | null
  reviewerId?: string
  reviewedAt?: string
  rejectReason?: string
  closedAt?: string
  timeline?: WorkOrderLogEntry[]
  createdBy?: string
  createdAt?: string
  updatedAt?: string
}

// ── Circuit ──────────────────────────────────────────────────────────────

export interface CircuitRequest {
  panelBoxDeviceId?: number | null
  circuitNumber: string
  circuitName?: string
  taipowerAccount?: string
  usageType?: string
}

export interface CircuitResponse {
  id: number
  panelBoxDeviceId?: number | null
  circuitNumber: string
  circuitName?: string
  taipowerAccount?: string
  usageType?: string
  status: string
  createdAt?: string
}

// ── Contract ─────────────────────────────────────────────────────────────

export interface ContractRequest {
  contractCode: string
  contractName: string
  budgetYear?: number | null
  procurementNumber?: string
  contractorName?: string
  contractorContact?: string
  assetCategory?: string
  quantity?: number | null
  startDate?: string | null
  endDate?: string | null
  acceptanceDate?: string | null
  warrantyYears?: number | null
}

export interface ContractResponse {
  id: number
  contractCode: string
  contractName: string
  budgetYear?: number | null
  procurementNumber?: string
  contractorName?: string
  contractorContact?: string
  quantity?: number | null
  startDate?: string | null
  endDate?: string | null
  acceptanceDate?: string | null
  warrantyExpiry?: string | null
  status: string
  createdBy?: string
  createdAt?: string
  updatedAt?: string
}

File: 01-cli/pkg/dto/device.go
package dto

type DeviceResponse struct {
	ID             int64            `json:"id"`
	Name           string           `json:"name,omitempty"`
	DeviceCode     string           `json:"deviceCode"`
	DeviceName     string           `json:"deviceName,omitempty"`
	DeviceType     string           `json:"deviceType"`
	Status         DeviceStatus     `json:"status"`
	SerialNumber   string           `json:"serialNumber,omitempty"`
	TenantID       string           `json:"tenantId,omitempty"`
	DeptID         *int64           `json:"deptId,omitempty"`
	DeptName       string           `json:"deptName,omitempty"`
	ContractID     *int64           `json:"contractId,omitempty"`
	ContractCode   string           `json:"contractCode,omitempty"`
	CircuitID      *int64           `json:"circuitId,omitempty"`
	CircuitNumber  string           `json:"circuitNumber,omitempty"`
	ParentDeviceID *int64           `json:"parentDeviceId,omitempty"`
	CreatedAt      string           `json:"createdAt,omitempty"`
	UpdatedAt      string           `json:"updatedAt,omitempty"`
	Children       []DeviceResponse `json:"children,omitempty"`
}

type DeviceRequest struct {
	Name             string         `json:"name,omitempty"`
	DeviceType       string         `json:"deviceType"`
	DeviceCode       string         `json:"deviceCode"`
	DeviceName       string         `json:"deviceName,omitempty"`
	SerialNumber     string         `json:"serialNumber,omitempty"`
	DeptID           *int64         `json:"deptId,omitempty"`
	ContractID       *int64         `json:"contractId,omitempty"`
	CircuitID        *int64         `json:"circuitId,omitempty"`
	ParentDeviceID   *int64         `json:"parentDeviceId,omitempty"`
	ConnectivityType string         `json:"connectivityType,omitempty"`
	InstalledAt      string         `json:"installedAt,omitempty"`
	Properties       map[string]any `json:"properties,omitempty"`
}

type DeviceStatsResponse struct {
	TotalByType    map[string]int64 `json:"totalByType,omitempty"`
	TotalByStatus  map[string]int64 `json:"totalByStatus,omitempty"`
	ByType         map[string]int64 `json:"byType"`
	ByStatus       map[string]int64 `json:"byStatus"`
	OnlineRate     float64          `json:"onlineRate"`
	OpenDispatches int64            `json:"openDispatches,omitempty"`
	OpenWorkOrders int64            `json:"openWorkOrders"`
}
